infer/inference.cc

```diff
@@ -86,6 +86,21 @@ unique_ptr<cfg::CFG> Inference::run(core::Context ctx, unique_ptr<cfg::CFG> cfg)
         }
         Environment &current = outEnvironments[bb->id];
         current.initializeBasicBlockArgs(*bb);
+
+        // We very much want to limit access to "global" data structures downstream.
+        // In particular, processBinding should only need to know about the current binding (nothing
+        // about the surrounding BasicBlock, and definitely nothing about BasicBlocks that don't
+        // contain the current binding being processed).
+        //
+        // That means that only here in inference do we know about the structure of the CFG.
+        // But we want to pluck some limited information about the parent block and pass it into
+        // processBinding, only so that some errors that would already be reported can be improved.
+        // parentUpdateKnowledgeReceiver is such a thing.
+        //
+        // In general we should carefully limit the set of information that processBinding requires,
+        // to make it easy to reason about correctness.
+        optional<cfg::BasicBlock::BlockExitCondInfo> parentUpdateKnowledgeReceiver;
```

Comment:
There are precious few comments in Sorbet explaining *why* things are done a certain way; thank you for adding this one.

I don't know that I am convinced that trying to limit what information we pass in is beneficial.  I can see the argument for it, but trying to avoid passing the necessary non-local information that we could use just in the error case means that we wind up computing this information in many many cases where it's never going to get used.  Have you tried benchmarking this change wrt typechecking performance?

Reply:
After a quick test, it seems that maybe this adds 0.29% time to pay-server master, which we think is basically within the margin of error of noise

from discussion offline: an alternative to avoid computing this except when needed might be to pass an opaque handle into `processBinding` that only exposes one helper method which computes the relevant information and is called only when we're sure that processBinding (or even dispatchCallSymbol) needs it. we thought that amounted to quite a lot of ceremony, all things considered.

we did both generally agree that it is a nice property of `processBinding` that it can't currently do things like lookup types of variables in parent blocks (i.e. in previous Environments), which makes it easier to reason about what can possibly affect how an individual binding typechecks or not, but we did think that there might arise future features which want to thread more information down into processBinding, at which point we might want to reconsider how we pass ad-hoc information down from inference.cc into environment.cc